### PR TITLE
feat: integrate react-grid-layout as atom

### LINKED ISSUE #2047: Integrate Read Grid Layout to platform as atom
## Description
- Current component tree is implemented as a component tree as:
- Parent Node
   - Children1 node
   - Children2 node (linked as an single linked list)

and we use CSS style to style the elements. We need an improvement and innovation for a new layout engine

The elements should be able to be resized, dragged, and drop using grid layout (Ref: https://github.com/react-grid-layout/react-grid-layout)

## Tracking
- [ ] POC layout engine
- [ ] TBC

### PR DESCRIPTION
**Pull request by tinhw3**
## Video or Image

https://user-images.githubusercontent.com/117329462/213129391-dffcd512-aa1a-45e6-839c-4357014ddefd.mov

## Related Issue(s)

Fixes #2047

### CODE REVIEW

File: libs/frontend/platform/atoms/src/custom/grid-layout/GridLayout.tsx [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+      <ResponsiveReactGridLayout
+        breakpoints={{ lg: 1200, md: 996, sm: 768, xs: 480, xxs: 0 }}
+        cols={{ lg: 12, md: 10, sm: 6, xs: 4, xxs: 2 }}
+        containerPadding={[0, 0]}
+        layouts={restProps.static ? previewLayouts : layouts}
+        margin={[0, 0]}
+        onLayoutChange={onLayoutChange}
+        rowHeight={30}
+        // eslint-disable-next-line react/jsx-props-no-spreading
+        {...restProps}
```
Comment (webberwang) [inline libs/frontend/platform/atoms/src/custom/grid-layout/GridLayout.tsx:89]: We should de-structure props and explicitly override the props with default values like `breakpoints` etc

File: libs/frontend/platform/atoms/src/custom/grid-layout/GridLayout.tsx [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        // TODO: handle primative situation if neccessary
+        if (!child || !React.isValidElement(child)) {
+          return null
+        }
+
+        return <div key={child.key}>{child}</div>
+      })
+    }, [children])
+
+    const layouts = restProps.layouts || {}
```
Comment (webberwang) [inline libs/frontend/platform/atoms/src/custom/grid-layout/GridLayout.tsx:36]: create `makeLayout` function

**Review by webberwang — CHANGES_REQUESTED** [review]

File: libs/frontend/platform/atoms/src/custom/grid-layout/GridLayout.tsx [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    const previewLayouts = restProps.layouts || {}
+
+    if (restProps.static) {
+      const keys = Object.keys(layouts)
+      keys.forEach((key) => {
+        layouts[key] = layouts[key]?.map((ele) => ({
+          ...ele,
+          static: restProps.static,
+        })) as Array<Layout>
```
Comment (webberwang) [inline libs/frontend/platform/atoms/src/custom/grid-layout/GridLayout.tsx:45]: could try https://www.npmjs.com/package/object-typed to preserve types so we may not have to cast

**Review by webberwang — CHANGES_REQUESTED** [review]

File: libs/frontend/platform/atoms/src/custom/grid-layout/GridLayout.tsx [hunk 4]
```diff
@@ -33,17 +34,50 @@ export const GridLayout = React.memo(
...
-        layouts[key] = layouts[key]?.map((ele) => ({
-          ...ele,
-          static: restProps.static,
-        })) as Array<Layout>
-      })
+      if (restProps.static) {
+        ObjectTyped.keys(layouts).forEach((key) => {
+          layouts[key] = layouts[key]?.map((ele) => ({
+            ...ele,
+            static: restProps.static,
```
Comment (webberwang) [inline libs/frontend/platform/atoms/src/custom/grid-layout/GridLayout.tsx:45]: cast earlier at the `Layout` level

File: libs/frontend/platform/atoms/src/custom/grid-layout/GridLayout.tsx [hunk 5]
```diff
@@ -33,17 +34,50 @@ export const GridLayout = React.memo(
...
+    }
+
+    // Make the RGL cols
+    const makeCols = () => {
+      const cols = { lg: 12, md: 10, sm: 6, xs: 4, xxs: 2 }
+
+      if (restProps.cols) {
+        ObjectTyped.keys(cols).forEach((key) => {
+          if (restProps.cols && restProps.cols[key]) {
+            cols[key] = restProps.cols[key]!
```
Comment (webberwang) [inline libs/frontend/platform/atoms/src/custom/grid-layout/GridLayout.tsx:75]: is `!` needed?

**Review by webberwang — CHANGES_REQUESTED** [review]

File: libs/frontend/platform/atoms/src/custom/grid-layout/GridLayout.tsx [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+export interface RenderedComponentProps {
+  [DATA_ELEMENT_ID]: string
+  ['static']: boolean
+}
+
+export const GridLayout = React.memo(
+  ({ children, ...restProps }: ResponsiveProps & RenderedComponentProps) => {
+    const elementId = restProps[DATA_ELEMENT_ID]
+    const isComponentMountedFirstTime = useRef(false)
```
Comment (webberwang) [inline libs/frontend/platform/atoms/src/custom/grid-layout/GridLayout.tsx:18]: is there a better way for this?

**Review by webberwang — CHANGES_REQUESTED** [review]

File: libs/frontend/platform/atoms/src/custom/grid-layout/GridLayout.tsx (hunk 5 above)
Comment (tinhw3) [reply libs/frontend/platform/atoms/src/custom/grid-layout/GridLayout.tsx:75]: so weird that vscode warning `restProps.cols[key]` is `number | undefined`

File: libs/frontend/platform/atoms/src/custom/grid-layout/GridLayout.tsx (hunk 6 above)
Comment (tinhw3) [reply libs/frontend/platform/atoms/src/custom/grid-layout/GridLayout.tsx:18]: you mean for this `isComponentMountedFirstTime`?

File: libs/frontend/platform/atoms/src/custom/grid-layout/GridLayout.tsx (hunk 4 above)
Comment (tinhw3) [reply libs/frontend/platform/atoms/src/custom/grid-layout/GridLayout.tsx:45]: Sorry @webberwang , I dont understand this one. Do you mean using `ele. static = restProps.static`?

File: libs/frontend/platform/atoms/src/custom/grid-layout/GridLayout.tsx (hunk 6 above)
Comment (webberwang) [reply libs/frontend/platform/atoms/src/custom/grid-layout/GridLayout.tsx:18]: right

File: libs/frontend/platform/atoms/src/custom/grid-layout/GridLayout.tsx (hunk 4 above)
Comment (webberwang) [reply libs/frontend/platform/atoms/src/custom/grid-layout/GridLayout.tsx:45]: I mean cast `as Layout` instead somewhere

**Pull request merged**